beluga/include/beluga/primitives.hpp

```diff
@@ -67,27 +67,27 @@ struct has_member_state : std::false_type {};
 template <class T>
 struct has_member_state<T, std::void_t<decltype(std::declval<T>().state())>> : std::true_type {};
 
-template <typename T>
+template <class T>
 inline constexpr bool has_member_state_v = has_member_state<T>::value;
 
-/// \endcond
+template <class T, class = void>
+struct has_non_member_state : std::false_type {};
 
-/// Default state implementation.
-/**
- * Assumes that the particle is a tuple and the first element is the state.
- */
-template <typename T>
-constexpr decltype(auto) state(T&& t) noexcept(noexcept(std::get<0>(std::forward<T>(t)))) {
-  return std::get<0>(std::forward<T>(t));
-}
+template <class T>
+struct has_non_member_state<T, std::void_t<decltype(state(std::declval<T>()))>> : std::true_type {};
+
+template <class T>
+inline constexpr bool has_non_member_state_v = has_non_member_state<T>::value;
+
+/// \endcond
 
 /// Customization point object type for accessing the `state` of a particle.
 /**
  * See https://en.cppreference.com/w/cpp/ranges/cpo.
  */
 struct state_fn {
   /// Overload for when the particle type defines a member method.
-  template <typename T, std::enable_if_t<has_member_state_v<T>, int> = 0>
+  template <class T, std::enable_if_t<has_member_state_v<T>, int> = 0>
   constexpr decltype(auto) operator()(T&& t) const noexcept(noexcept(std::forward<T>(t).state())) {
```

Comment:
@nahueespinosa TIL building `noexcept` specifiers using `noexcept` operators is a thing.